OpenRA.Mods.Common/Traits/ProductionParadrop.cs

```diff
@@ -0,0 +1,170 @@
+			var startPos = self.Location + new CVec(owner.World.Map.Bounds.Width, 0);
+			var endPos = new CPos(owner.World.Map.Bounds.Left - 5, self.Location.Y);
+
+			// Assume a single exit point for simplicity
+			var exit = self.Info.TraitInfos<ExitInfo>().First();
+
+			foreach (var tower in self.TraitsImplementing<INotifyDelivery>())
+				tower.IncomingDelivery(self);
+
+			var info = (ProductionParadropInfo)Info;
+			var actorType = info.ActorType;
+
+			owner.World.AddFrameEndTask(w =>
+				{
+					if (!self.IsInWorld || self.IsDead)
+						return;
+
+					var altitude = self.World.Map.Rules.Actors[actorType].TraitInfo<AircraftInfo>().CruiseAltitude;
+					var actor = w.CreateActor(actorType, new TypeDictionary
+						{
+							new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WDist.Zero, WDist.Zero, altitude)),
+							new OwnerInit(owner),
+							new FacingInit(64)
+						});
+
+					actor.QueueActivity(new Fly(actor, Target.FromCell(w, self.Location + new CVec(0, 0))));
+					actor.QueueActivity(new CallFunc(() =>
+						{
+							if (!self.IsInWorld || self.IsDead)
+								return;
+
+							foreach (var cargo in self.TraitsImplementing<INotifyDelivery>())
+								cargo.Delivered(self);
+							
+							self.World.AddFrameEndTask(ww => DoProduction(self, producee, exit, factionVariant));
+							Game.Sound.PlayNotification(self.World.Map.Rules, self.Owner, "Speech", info.ReadyAudio, self.Owner.Faction.InternalName);
+						}));
+
+					actor.QueueActivity(new Fly(actor, Target.FromCell(w, endPos)));
+					actor.QueueActivity(new RemoveSelf());
+				});
+
+			return true;
+		}
+
+		public override void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, string factionVariant)
+		{
+			var exit = CPos.Zero;
+			var exitLocation = CPos.Zero;
+			var target = Target.Invalid;
+
+			var info = (ProductionParadropInfo)Info;
+			var actorType = info.ActorType;
+
+			var bi = producee.TraitInfoOrDefault<BuildableInfo>();
+			if (bi != null && bi.ForceFaction != null)
+				factionVariant = bi.ForceFaction;
+
+			var altitude = self.World.Map.Rules.Actors[actorType].TraitInfo<AircraftInfo>().CruiseAltitude;
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+			};
+
+			if (self.OccupiesSpace != null)
+			{
+				exit = self.Location + exitinfo.ExitCell;
+				var spawn = self.CenterPosition + exitinfo.SpawnOffset + new WVec(WDist.Zero, WDist.Zero, altitude);
+				var to = self.World.Map.CenterOfCell(exit);
+
+				var fi = producee.TraitInfoOrDefault<IFacingInfo>();
+				var initialFacing = exitinfo.Facing < 0 ? Util.GetFacing(to - spawn, fi == null ? 0 : fi.GetInitialFacing()) : exitinfo.Facing;
+
+				exitLocation = rp.Value != null ? rp.Value.Location : exit;
+				target = Target.FromCell(self.World, exitLocation);
+
+				td.Add(new LocationInit(exit));
+				td.Add(new CenterPositionInit(spawn));
+				td.Add(new FacingInit(initialFacing));
+			}
+
+			self.World.AddFrameEndTask(w =>
+				{
```

Comment:
This block should be unindented by one tab to line up with the rest of the outer block.